akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/AccountExampleTest.java

```diff
@@ -0,0 +1,149 @@
+    ref.tell(new CreateAccount(probe.getRef()));
+    probe.expectMessage(Confirmed.INSTANCE);
+    ref.tell(new Deposit(BigDecimal.valueOf(100), probe.getRef()));
+    probe.expectMessage(Confirmed.INSTANCE);
+    ref.tell(new Withdraw(BigDecimal.valueOf(10), probe.getRef()));
+    probe.expectMessage(Confirmed.INSTANCE);
+  }
+
+  @Test
+  public void rejectWithdrawOverdraft() {
+    EntityRef<AccountCommand> ref = sharding().entityRefFor(AccountEntity.ENTITY_TYPE_KEY, "3");
+    TestProbe<OperationResult> probe = testKit.createTestProbe(OperationResult.class);
+    ref.tell(new CreateAccount(probe.getRef()));
+    probe.expectMessage(Confirmed.INSTANCE);
+    ref.tell(new Deposit(BigDecimal.valueOf(100), probe.getRef()));
+    probe.expectMessage(Confirmed.INSTANCE);
+    ref.tell(new Withdraw(BigDecimal.valueOf(110), probe.getRef()));
+    probe.expectMessageClass(Rejected.class);
+  }
+
+  @Test
+  public void handleGetBalance() {
+    EntityRef<AccountCommand> ref = sharding().entityRefFor(AccountEntity.ENTITY_TYPE_KEY, "4");
+    TestProbe<OperationResult> opProbe = testKit.createTestProbe(OperationResult.class);
+    ref.tell(new CreateAccount(opProbe.getRef()));
+    opProbe.expectMessage(Confirmed.INSTANCE);
+    ref.tell(new Deposit(BigDecimal.valueOf(100), opProbe.getRef()));
+    opProbe.expectMessage(Confirmed.INSTANCE);
+
+    TestProbe<CurrentBalance> getProbe = testKit.createTestProbe(CurrentBalance.class);
+    ref.tell(new GetBalance(getProbe.getRef()));
+    assertEquals(
+        BigDecimal.valueOf(100), getProbe.expectMessageClass(CurrentBalance.class).balance);
+  }
+
+  @Test
+  public void beUsableWithAsk() throws Exception {
+    Timeout timeout = Timeout.create(Duration.ofSeconds(3));
+    EntityRef<AccountCommand> ref = sharding().entityRefFor(AccountEntity.ENTITY_TYPE_KEY, "5");
+    CompletionStage<OperationResult> createResult =
+        ref.ask(replyTo -> new CreateAccount(replyTo), timeout);
+    assertEquals(Confirmed.INSTANCE, createResult.toCompletableFuture().get(3, TimeUnit.SECONDS));
+
+    // above works because then the response type is inferred by the lhs type
+    // below requires (ActorRef<OperationResult> replyTo)
+
+    assertEquals(
+        Confirmed.INSTANCE,
+        ref.ask(
+                (ActorRef<OperationResult> replyTo) ->
+                    new Deposit(BigDecimal.valueOf(100), replyTo),
+                timeout)
+            .toCompletableFuture()
+            .get(3, TimeUnit.SECONDS));
+
+    assertEquals(
+        Confirmed.INSTANCE,
+        ref.ask(
+                (ActorRef<OperationResult> replyTo) ->
+                    new Withdraw(BigDecimal.valueOf(10), replyTo),
+                timeout)
+            .toCompletableFuture()
+            .get(3, TimeUnit.SECONDS));
+
+    BigDecimal balance =
+        ref.ask((ActorRef<CurrentBalance> replyTo) -> new GetBalance(replyTo), timeout)
+            .thenApply(currentBalance -> currentBalance.balance)
```

Comment:
Note that the explicit `(ActorRef<CurrentBalance> replyTo)` type is required here. Alternatively it has to be split up in two steps
```
CompletionStage<OperationResult> result = .ask(replyTo -> new GetBalance(replyTo), timeout)
result.thenApply(currentBalance -> currentBalance.balance)
```

@renatocaval Would this be ok in Lagom?

An alternative solution would be to add another `Class<Res> responseClass` parameter to the ask signature:
```
        ref.ask(OperationResult.class, replyTo -> new GetBalance(replyTo), timeout)
	            .thenApply(currentBalance -> currentBalance.balance)
```
but that isn't so nice for when one would like to split it up
```
CompletionStage<OperationResult> result = ref.ask(OperationResult.class, replyTo -> new GetBalance(replyTo), timeout)
```

Reply:
I think the explicit parameter type for the lambda is better than two-step or an explicit `Class` parameter.

Reply:
For Lagom that means we go from:

```java
ref.ask(GetBalance)
   .thenApply(currentBalance -> currentBalance.balance);
```

to 

```java
 ref.ask((ActorRef<CurrentBalance> replyTo) -> new GetBalance(replyTo), timeout)
    .thenApply(currentBalance -> currentBalance.balance)
```

From an Akka user perspective, there is increased boilerplate with increased value. It's now typed. 

From a Lagom user perspective, there is only increased boilerplate. They have already a typed API. 

Unfortunately, I guess that there is not much we can do here. 

From the two proposals, I prefer the one using a simple lambda. We are not removing much friction in passing a `Class`.

Reply:
A few more alternatives that compile:

```scala
 BigDecimal balance4 =
        ref.ask(AccountExampleTest::blah, timeout)
            .thenApply(currentBalance -> currentBalance.balance)
            .toCompletableFuture()
            .get(3, TimeUnit.SECONDS);

    BigDecimal balance3 =
        ref.<CurrentBalance>ask(replyTo -> new GetBalance(replyTo), timeout)
            .thenApply(currentBalance -> currentBalance.balance)
            .toCompletableFuture()
            .get(3, TimeUnit.SECONDS);

    BigDecimal balance2 =
        ref.ask(GetBalance::new, timeout)
            .thenApply(currentBalance -> currentBalance.balance)
            .toCompletableFuture()
            .get(3, TimeUnit.SECONDS);
    assertEquals(BigDecimal.valueOf(90), balance2);
  }

  private static GetBalance blah(ActorRef<CurrentBalance> replyTo) {
    return new GetBalance(replyTo);
  }
```

So method references, including ctr, work. Java just can't infer types from usage in a lambda, even a simple one.